proposals/field-keyword.md

```diff
@@ -230,74 +230,110 @@ public struct S
+- Patterns that use the `field` keyword should feel as though they've always been part of the language. Avoid making the user jump through hoops to enable Nullable Reference Types in code that is perfectly idiomatic for the `field` keyword feature.
 
-    public string P { get => field; set => field = value; }
-}
-```
+One of the key scenarios is lazily initialized properties:
 
 ```cs
-class C
+public class C
 {
-    public string P { get => field; set => field = value; } = "";
+    public C() { } // It would be undesirable to warn about 'Prop' being uninitialized here
+
+    string Prop => field ??= GetPropValue();
 }
 ```
 
-#### Open question: nullability of `field`
+The following nullability rules will apply not just to properties that use the `field` keyword, but also to existing auto properties.
 
-In the same vein as how `var` infers as nullable for reference types, the `field` type is the nullable type of the property whenever the property's type is not a value type. Otherwise, `field ??` would appear to be followed by dead code, and it avoids producing a misleading warning in the following example:
+#### Terms
 
+A property is a *field-backed property* if it meets any of the following conditions:
+1. It contains any automatically implemented accessors (`get;`/`set;`/`init;`).
+2. It uses the `field` keyword within any accessor bodies.
+
+Some examples of *field-backed properties* include:
 ```cs
-public string AmbientValue
-{
-    get => field ?? parent.AmbientValue;
-    set
-    {
-        if (value == parent.AmbientValue)
-            field = null; // No warning here. Resume following the parent's value.
-        else
-            field = value; // Stop following the parent's value
-    }
-}
+public string Prop1 { get; set; }
+public string Prop2 { get => field; set; }
+public string? Prop3 { get => field; }
 ```
 
-`var` was designed to declare nullability so that subsequent assignments to the variable could be nullable, due to established patterns in C#. It's expected that the same rationale would apply to property backing fields.
+The variable denoted by the `field` keyword in a property's accessors is the *backing field* of that property.
+
+#### Nullability of the *backing field*
+
+The *backing field* has the same type as the property. However, its nullable annotation may differ from the property. To determine this nullable annotation, we introduce the concept of *null-resilience*. *Null-resilience* intuitively means that the property's `get` accessor preserves null-safety even when the field contains the `default` value for its type.
+
+A *field-backed property* is determined to be *null-resilient* or not by performing a special nullable analysis of its `get` accessor.
+- For the purposes of this analysis, `field` is temporarily assumed to have *annotated* nullability, e.g. `string?`. This causes `field` to have *maybe-null* or *maybe-default* initial state in the `get` accessor, depending on its type.
```

Comment:
@RikkiGibson For an `int` property, does "annotated nullability" mean `int?`- that is, to literally add a syntactic `?`- or is it defined to mean "maybe-default"?

I'm curious in case we would want to explicitly say anything about value types remaining unaffected by this. Similar question lower down; would we want an explicit rule that bails out of "The nullability of the backing field is determined as follows:" like we do for if `[field: MaybeNull]` is specified?

Reply:
Nullable annotation is not a thing for value types. In that case, a question mark is a `System.Nullable` (reified) type rather than merely a type annotation.

*maybe-default* is only really used for type parameters. In that case, the meanings of the flow states are:
- maybe-default: the value may be null, even if `T` is non-nullable reference type.
- maybe-null: the value may be null, only if `T` itself permits null (e.g. it is ok to return this value when the return type is `T`, but it is not ok to dereference it.)
- not-null: same as usual.

Reply:
Does that mean that when considering a value-typed property, it goes without saying that you stop after the first sentence and consider everything else irrelevant?

> The *backing field* has the same type as the property. However, its nullable annotation may differ from the property. To determine this nullable annotation, we introduce the concept of *null-resilience*.

Reply:
Yes exactly.